[Low Bing Heng] iP

BuffWuff1712 opened the pull request:
### BingBong

> “Empower your productivity, one task at a time.”

BingBong is your go-to companion for staying on top of your to-dos, offering a seamless experience that's:

- **User-friendly**: Intuitive and designed with simplicity in mind.
- **Lightweight**: No unnecessary frills—just what you need to get things done.
- **Blazingly Efficient**: Lightning-fast, so you can focus on your tasks, not on using the tool.

Here's how easy it is to get started with BingBong:

1. Download it from [this link](https://github.com/BuffWuff1712/ip).
2. Run the application—no complicated setup needed.
3. Add your tasks effortlessly.
4. Let BingBong handle the rest—track your progress with a smile 😊.

And the best part? It’s completely **FREE!**

Key Features:

- [ ] Task Management: Keep your tasks organized and accessible.
- [ ] Deadline Tracking: (coming soon) Never miss a deadline again.
- [ ] Smart Reminders: (coming soon) Stay on top of your schedule with timely nudges.


If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

BuffWuff1712 committed:
Level-0

BuffWuff1712 committed:
Added Bot Logic and UI

BuffWuff1712 committed:
Add and List features

BuffWuff1712 committed:
Added mark and unmark tasks feature

BuffWuff1712 committed:
Added Todo, Deadline and Event task types

BuffWuff1712 committed:
added automated I/O tests

BuffWuff1712 committed:
Added exception handling, and BingBongException class

BuffWuff1712 committed:
Added delete task feature

BuffWuff1712 committed:
added enum CommandType

BuffWuff1712 committed:
Add saving to and loading from a text file

BuffWuff1712 committed:
Merge branch 'branch-Level-7'
This merge incorporates the changes from the feature-branch into the master branch. The feature-branch includes the following updates:

- Added functionality to save tasks to a text file automatically whenever the task list is modified.
- Implemented loading of tasks from the text file when the chatbot starts, ensuring persistence of tasks across sessions.
- Handled cases where the data file or directory does not exist, creating them as needed.

This merge also maintains the commit history from the feature branch,
allowing for better traceability of the implemented changes.

BuffWuff1712 committed:
Add DateTimeHandler for date and time recognition

BuffWuff1712 committed:
Add command to list events/deadlines for specific dates

BuffWuff1712 committed:
Merge branch 'branch-Level-8'
Add support for date and time handling in tasks (Level 8)

This update introduces the ability for BingBongBot to parse, format, and manage dates and times associated with tasks, specifically for deadlines and events. The enhancements include:

- User-Friendly Date Input: Users can now enter dates and times in a standard format (e.g., `d/M/yyyy HHmm`), which the bot will interpret and store correctly.
- Enhanced Task Management: By understanding dates and times, BingBongBot can now accurately track when tasks are due and when events are scheduled, providing a more organized and efficient task management experience.
- Task Filtering by Date: The bot can list tasks occurring on a specific date, allowing users to quickly view all deadlines and events scheduled for that day.
- Consistency Across Sessions: Dates and times are saved and loaded in a consistent format, ensuring that task deadlines and event times are preserved accurately between sessions.

BuffWuff1712 committed:
Add Parser and TaskList classes

BuffWuff1712 committed:
Add Command classes

BuffWuff1712 committed:
Merge branch 'A-MoreOOP'
Refactor code to improve Object-Oriented Design by extracting Parser, Command, and TaskList classes

This commit refines the structure of the codebase by introducing several new classes, enhancing the overall Object-Oriented Design:

- **Extracted Parser Class**:
  - Moved the command parsing logic from the `BingBongBot` class into a dedicated `Parser` class.
  - This improves modularity and makes the command interpretation process more isolated and easier to maintain.

- **Introduced Command Classes**:
  - Created a set of `Command` classes (e.g., `AddCommand`, `ListCommand`, `MarkCommand`) to encapsulate different types of user commands.
  - This follows the Command design pattern, promoting flexibility and making it easier to add or modify commands in the future without disrupting the bot’s core logic.

- **Extracted TaskList Class**:
  - Introduced a `TaskList` class to encapsulate the collection of tasks.
  - This class manages tasks more effectively, providing a clear API for adding, removing, and retrieving tasks.
  - By abstracting the task management into its own class, the bot's code becomes cleaner and easier to manage.

- **Enhanced Readability and Maintainability**:
  - The refactor breaks down the monolithic logic into smaller, well-defined classes, improving code readability and adherence to the Single Responsibility Principle.
  - Each class now has a focused responsibility, making the code easier to understand and maintain.

- **Improved Flexibility for Future Enhancements**:
  - The new architecture supports easier extension of functionality, such as adding new commands or modifying task management, with minimal impact on the overall system.

This refactor represents a significant step towards a more scalable, flexible, and maintainable codebase, setting the stage for future development and enhancements.

BuffWuff1712 committed:
Add Packages

BuffWuff1712 committed:
Merge branch 'branch-A-Packages'
Organised the classes into suitable java packages for better management
of java files

BuffWuff1712 committed:
Merge branch 'add-gradle-support'
Adding gradle support for Automation Testing

BuffWuff1712 committed:
Add ParserTest and TodoTest JUnit test files

BuffWuff1712 committed:
Merge branch 'branch-A-JUnit'
Added basic JUnit tests to test the functionality of Parser and Todo
classes

BuffWuff1712 committed:
Add main.jar file

BuffWuff1712 committed:
Merge branch 'branch-A-Jar'

BuffWuff1712 committed:
Add JavaDoc to core and task package files

BuffWuff1712 committed:
Edit code to comply with Coding Standard

BuffWuff1712 committed:
Add Find command using keyword

BuffWuff1712 committed:
Merge branch 'branch-Level-9'
Master branch is to be updated with the new Find feature

BuffWuff1712 committed:
Merge branch 'branch-A-CodingStandard'
Master branch now complies with coding standard after merge

BuffWuff1712 committed:
Merge branch 'branch-A-JavaDoc'

Fixed merge conflicts between master and branch-A-JavaDoc

BuffWuff1712 committed:
Add CheckStyle config files

BuffWuff1712 committed:
Merge branch 'branch-A-CheckStyle'
Added CheckStyle files to enforce consistent coding standards and detect style violations across the project. This merge ensures early detection of issues and promotes maintainable, high-quality code.

BuffWuff1712 committed:
Add Basic JavaFX GUI to BingBong app

BuffWuff1712 committed:
Merge branch 'branch-Level-10'
Add a basic JavaFX GUI to the chatbot application

This merge integrates a JavaFX-based graphical user interface (GUI) into the chatbot, replacing the previous text-based interaction model. The GUI provides a more user-friendly and visually appealing way for users to interact with the chatbot, improving the overall user experience.

Key changes include:
- Implementation of the `MainWindow` class to manage the main application window.
- Addition of the `DialogBox` class to format and display user and bot messages within the GUI.
- Integration of the `BingBongBot` class with the GUI to handle user commands and responses.
- Setup of FXML files and resources for the layout and design of the interface.

This merge is necessary to transition the chatbot from a console-based application to a modern, visually intuitive GUI application, making it more accessible and engaging for users.

BuffWuff1712 committed:
Tidy up codebase to follow coding standards

BuffWuff1712 committed:
Merge branch 'branch-A-Varargs'
Update codebase to comply with coding standards

This merge was necessary to ensure that the entire codebase adheres to established coding standards, improving overall readability, maintainability, and consistency across the project. Key changes include:

- Reordering import statements to follow the proper grouping and sequence.
- Breaking down long lines of code and JavaDocs to comply with the line length limit.
- Refactoring method signatures and variable names to adhere to naming conventions.
- Adding or modifying JavaDocs to ensure that all classes and methods are properly documented.

By enforcing these coding standards, we aim to make the code easier to understand and maintain for all current and future contributors.

Chronoxy commented on src/main/java/myapp/core/Storage.java:
```diff
@@ -0,0 +1,106 @@
+            String line;
+            while (reader.hasNext()) {
+                line = reader.nextLine();
+                String[] taskData = line.split(" \\| ");
```
I like how you used .split in your code. It makes it a lot 'cleaner'

Chronoxy commented on src/main/java/myapp/task/TaskList.java:
```diff
@@ -0,0 +1,98 @@
+        return taskList.remove(index);
+    }
+
+    /**
+     * Checks if the list is empty.
+     *
+     * @return {@code true} if the list contains no tasks, {@code false} otherwise.
+     */
+    public boolean isEmpty() {
+        return taskList.isEmpty();
+    }
+
+    /**
+     * Returns the task at the specified index.
+     *
+     * @param i the index of the task to retrieve.
+     * @return the {@link Task} at the specified index.
+     */
+    public Task get(int i) {
+        return taskList.get(i);
+    }
+
+    /**
+     * Returns the index of the first occurrence of the specified element in the list, or -1 if the list
+     *      does not contain the element.
+     *
+     * @param o the element to search for.
+     * @return the index of the first occurrence of the specified element, or -1 if not found.
+     */
+    public int indexOf(Object o) {
+        return taskList.indexOf(o);
+    }
+
+    /**
+     * Returns the number of tasks in the list.
+     *
+     * @return the number of tasks in the list.
+     */
+    public int size() {
+        return taskList.size();
```
May I ask why you chose to do it this way instead of just using it directly when called?

BuffWuff1712 replied:
Creating a TaskList class is a requirement.

Chronoxy replied:
I realized i am trolling with this comment 😅. Apologies man!

Chronoxy reviewed (commented):
All in all, I like how your code is very clean and self explanatory 😄

wassilililily commented on src/main/java/myapp/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,47 @@
+ */
+public class DeadlineCommand extends AddCommand {
+    /** The date and time by which the task should be completed. */
+    private LocalDateTime by;
```
Maybe instead of using prepositions as variable names, you can consider being more specific with the names such as 'deadlineTime'

wassilililily commented on src/main/java/myapp/task/Event.java:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @return the {@link LocalDateTime} when the event starts.
+     */
+    public LocalDateTime getFrom() {
```
Ending your method name with a preposition might confuse some people, maybe you can try naming the method to be more explicit?

wassilililily commented on src/main/java/myapp/task/Task.java:
```diff
@@ -0,0 +1,71 @@
+     *
+     * @return "X" if the task is done, otherwise a space (" ").
+     */
+    public String getStatusIcon() {
```
I like how clear the naming of your getter functions is! You just need to tidy up some of the variable and method names.

wassilililily commented on src/main/java/myapp/command/FindCommand.java:
```diff
@@ -0,0 +1,67 @@
+        boolean hasTasks = false;
+        int count = 1;
+
+        for (Task task : tasks) {
```
I like how you pluralise the name of the multi-valued variable!

wassilililily reviewed (commented):
Very clear and consistent names for methods and variables!
Just need to be aware of naming/ending your variables and methods with a preposition as it might be confusing to others.

BuffWuff1712 committed:
Add assertions to important parts of the codebase

BuffWuff1712 committed:
Refactor code to improve readability
Refactored several methods in the project to adhere to the Single Level of
Abstraction Principle (SLAP). This improves code readability, maintainability,
and testability by ensuring that each method is responsible for one level of
abstraction only.

Changes include:
- Extracted smaller, focused methods from large functions that were doing
  too much at once.
- Renamed methods to better reflect their responsibilities and simplify
  understanding.
- Moved repeated logic into helper functions to reduce duplication.
- Improved overall separation of concerns across several classes.

Why:
- SLAP promotes cleaner code with more modular functions, allowing easier
  testing and future modifications.
- The previous structure had large methods doing multiple tasks, making it
  harder to debug and extend.
- Following SLAP enhances maintainability as each function now has a
  well-defined role.

BuffWuff1712 committed:
Refactor Storage class to use Java Streams
Refactored the load and save methods in the Storage class to make use of
Java Streams. This improves code readability, efficiency, and maintains
a more functional programming approach to task handling.

Changes:
- The load method now uses Files.lines() to read all lines from the file
  as a stream, and each line is processed and mapped to a Task object
  using a helper method.
- The save method now streams the tasks, converting each task to its file
  format and writing them to the file in a more concise manner.
- Added helper methods for task parsing and writing to simplify the main
  logic and keep it clean.

Why:
- Using streams allows for more declarative, readable, and maintainable
  code by processing tasks in a functional style.
- Improves performance for handling large files by leveraging lazy
  evaluation with streams.
- Reduces duplication and keeps the load and save logic more focused
  on their respective tasks.

BuffWuff1712 committed:
Merge pull request #1 from BuffWuff1712/branch-A-Assertions

Add assertions to important parts of the codebase

BuffWuff1712 committed:
Merge branch 'master' into branch-A-CodeQuality

Updating the CodeQuality branch with the latest master branch

BuffWuff1712 committed:
Merge branch 'master' into branch-A-Streams

Updating Streams branch with the latest master branch code.

BuffWuff1712 committed:
Merge pull request #2 from BuffWuff1712/branch-A-CodeQuality

 Merge branch-A-CodeQuality into master

BuffWuff1712 committed:
Merge branch 'master' into branch-A-Streams

Updating Streams branch with latest master code.

BuffWuff1712 committed:
Merge pull request #3 from BuffWuff1712/branch-A-Streams

This request is to confirm the merge of the streams branch into the master branch after all review comments have been addressed and tests have successfully passed. The refactor using Java Streams has been thoroughly reviewed and tested, ensuring it adheres to the goals of improving performance, readability, and maintainability.

BuffWuff1712 committed:
Remove stream in load method of Storage

BuffWuff1712 committed:
Add gradle.yml file to set up CI

BuffWuff1712 committed:
Merge branch 'branch-A-CI'
Add gradle.yml file for CI configuration

This merge integrates the CI configuration into the main branch using a `gradle.yml` file. The configuration sets up GitHub Actions to automate the build, test, and deployment processes for the project.

Key changes:
- Added `gradle.yml` to define the CI pipeline using GitHub Actions.
- Configures Gradle to run automated build and test jobs on each push and pull request.
- Ensures compatibility with Java 11 (or the specific Java version used in the project).
- Executes tasks like `clean`, `build`, and `test` to maintain code quality.

This addition enhances development workflows by automating testing and ensuring that builds are consistent across environments.

No conflicts with existing code.

BuffWuff1712 committed:
Add FixedDurationTask to possible task types

BuffWuff1712 committed:
Add HelpCommand for users to view possible commands

BuffWuff1712 committed:
Merge branch 'branch-B-FixedDurationTasks'
- Introduced `FixedDuration` task type, allowing users to create tasks with a set duration but no fixed start or end times.
- Command for adding a fixed duration task: `fixed <task> /period <timeInHours>`.
- Duration is stored in minutes, supporting both hours and minutes for flexibility.
- Updated toString and file storage to handle FixedDuration tasks.

- Added `HelpCommand` to display a list of all available commands.
- Grouped commands by functionality (task creation, viewing, management, search, exit).
- Each command includes usage examples and correct formatting for user guidance.

This merge improves task flexibility and enhances the user experience by providing helpful documentation.

BuffWuff1712 committed:
Add User Guide to project for users to reference

BuffWuff1712 committed:
Add unique responses to each command

BuffWuff1712 committed:
Merge branch 'branch-A-Personality'

BuffWuff1712 committed:
Add new exception classes and more error handling

BuffWuff1712 committed:
Merge branch 'branch-A-MoreErrorHandling'

BuffWuff1712 committed:
Add new css styles and improve GUI

BuffWuff1712 committed:
Merge branch 'branch-A-BetterGui'

BuffWuff1712 committed:
Update README.md

BuffWuff1712 committed:
Resolve merge conflicts

BuffWuff1712 committed:
Edit app image link in README

BuffWuff1712 committed:
Add new test cases for Storage, Parser and DateTimeHandler

BuffWuff1712 committed:
Merge branch 'branch-A-MoreTesting'

BuffWuff1712 committed:
Remove tasks.txt and add delay to app close after bye command

BuffWuff1712 committed:
Generate JAR file with Java 17 and minor comment edits

BuffWuff1712 committed:
Add trim method to input to remove trailing whitespace

BuffWuff1712 committed:
Add error handling of deadline being earlier than current datetime

BuffWuff1712 committed:
Add toLowerCase method to ensure command comparison is case-insensitive

BuffWuff1712 committed:
Update Parser test cases
